k2/csrc/reverse.cu

```diff
@@ -0,0 +1,363 @@
+   reversed FSA, so an "offset" array is maintained to compute the 'idx01'.
+   Bear in mind, the additional final state for each reversed FSA doesn't have
+   a corresponding state in original FSA. 
+   */
+  Array1<int32_t> ConstructStateMapping() {
+    int32_t num_states = fsas_.TotSize(1);
+    Array1<int32_t> old2new_map(c_, num_states);
+    int32_t *old2new_map_data = old2new_map.Data();
+
+    Array1<int32_t> offset_idx01 = ExclusiveSum(non_empty_fsas_);
+    const int32_t *offset_idx01_data = offset_idx01.Data(),
+                  *fsas_row_ids1_data = fsas_.RowIds(1).Data(),
+                  *fsas_row_splits1_data = fsas_.RowSplits(1).Data(),
+
+    K2_EVAL(
+        c_, num_states, lambda_set_old2new_map, (int32_t idx01)->void {
+          // The 'idx0' must corresponds to a non-empty FSA as it is computed
+          // from the row_ids.
+          int32_t idx0 = fsas_row_ids1_data[idx01],
+                  idx0x = fsas_row_splits1_data[idx0],
+                  idx0x_final_state = fsas_row_splits1_data[idx0 + 1] - 1;
+          // Initialize as the offset.
+          int32_t new_idx01 = offset_idx01_data[idx0];
+          if (idx01 == idx0x) {
+            new_idx01 += idx0x_final_state;
+          } else if (idx01 == idx0x_final_state) {
+            new_idx01 += idx0x;
+          } else {
+            new_idx01 += idx01;
+          }
+          old2new_map_data[idx01] = new_idx01
+        });
+    return old2new_map;
+  }
+
+  
+  /*
+   Return the RaggedShape of the state-level reversed FsaVec.
+   Compared with the original FsaVec, we need to add an extra final state for
+   each non-empty FSA.
+   */
+  RaggedShape CreateReversedStateShape() {
+    Array1<int32_t> acc_non_empty_fsas = ExclusiveSum(non_empty_fsas_);
+    Array1<int32_t> reversed_row_split1 =
+      Plus(acc_non_empty_fsas, fsas_.RowSplits(1));
+    return RaggedShape2(&reversed_row_split1, nullptr, -1);
+  }
+
+
+  /*
+   Retrun the RaggedShape of the arc-level reversed FsaVec.
+   We generate the RaggedShape by figuring out the number of out-going arcs for
+   each reversed state. As the original source state will become to destination
+   state. So we collect the information by 'incoming_arcs'.
+
+   All cases will be one-to-one mapping, except that we need to add an extra
+   'final_arc' to the penultimate state of the reversed FSA (i.e. the
+   corresponding start state in original FSA).
+   */
+  RaggedShape CreateReversedArcSahpe(Ragged<int32_t> &incoming_arcs,
+                                     RaggedShape &reversed_states_shape) {
+    int32_t num_ori_states = fsas_.TotSize(1);
+    const int32_t *fsas_row_splits1_data = fsas_.RowSplits(1).Data(),
+                  *fsas_row_ids1_data = fsas_.RowIds(1).Data(), 
+                  *incoming_arcs_row_splits2_data =
+                    incoming_arcs.RowSplits(2).Data();
+
+    Array1<int32_t> num_arcs_per_reversed_state(
+        c_, reversed_states_shape.NumElements() + 1, 0);
```

Comment:
Do we have to initialize it to 0?

Reply:
After reversing, there is an extra new final state which has no outgoing arcs. So I initialize it to 0. Otherwise, I need another "K2_EVAL" to set it. So I choose a convenient way.